### PR TITLE
cleanup: consolidate validation and application of transaction inputs and outputs

### PR DESCRIPTION
**Pull request by joschisan**
The reduction in computational load from splitting the functions is negligible.

### CODE REVIEW

File: fedimint-server/src/net/api.rs [hunk 1]
```diff
@@ -103,39 +104,43 @@ impl ConsensusApi {
...
-            pub_keys.push(meta.pub_keys);
             funding_verifier.add_input(meta.amount);
+            public_keys.push(meta.pub_keys);
         }
-        transaction.validate_signature(pub_keys.into_iter().flatten())?;
 
-        for output in &transaction.outputs {
+        transaction.validate_signature(public_keys.into_iter().flatten())?;
+
+        for (output, out_idx) in transaction.outputs.iter().zip(0u64..) {
```
[the remark this answers is not in the document]
Comment (joschisan) [reply fedimint-server/src/net/api.rs:133]: Yes but then I could not write 
```rust
OutPoint { txid, out_idx }
```
directly.

File: modules/fedimint-wallet-server/src/lib.rs [hunk 2]
```diff
@@ -445,23 +445,7 @@ impl ServerModule for Wallet {
...
-        dbtx: &mut ModuleDatabaseTransaction<'c>,
-        input: &'b WalletInput,
-        cache: &Self::VerificationCache,
-    ) -> Result<InputMeta, ModuleError> {
-        let meta = self.validate_input(dbtx, cache, input).await?;
-        debug!(outpoint = %input.outpoint(), amount = %meta.amount.amount, "Claiming peg-in");
+        debug!(outpoint = %input.outpoint(), "Claiming peg-in");
 
         dbtx.insert_new_entry(
             &UTXOKey(input.outpoint()),
```
[the remark this answers is not in the document]
Comment (joschisan) [reply modules/fedimint-wallet-server/src/lib.rs]: Done

File: fedimint-server/src/consensus/mod.rs [hunk 3]
```diff
@@ -88,8 +88,8 @@ pub struct FedimintConsensus {
 }
 
 #[derive(Debug)]
-struct VerificationCaches {
-    caches: HashMap<ModuleInstanceId, DynVerificationCache>,
+pub struct VerificationCaches {
+    pub caches: HashMap<ModuleInstanceId, DynVerificationCache>,
```
[the remark this answers is not in the document]
Comment (joschisan) [reply fedimint-server/src/consensus/mod.rs:92]: Yep thats a little messy. Wanted to squash booth commits later anyways.

File: modules/fedimint-ln-server/src/lib.rs [hunk 4]
```diff
@@ -594,32 +595,24 @@ impl ServerModule for Lightning {
...
-        cache: &Self::VerificationCache,
-    ) -> Result<InputMeta, ModuleError> {
-        let meta = self.validate_input(dbtx, cache, input).await?;
-
         let account_db_key = ContractKey(input.contract_id);
+
         let mut contract_account = dbtx
             .get_value(&account_db_key)
             .await
             .expect("Should fail validation if contract account doesn't exist");
```
[the remark this answers is not in the document]
Comment (joschisan) [reply modules/fedimint-ln-server/src/lib.rs]: Done

File: fedimint-server/src/net/api.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (elsirion) [reply fedimint-server/src/net/api.rs:133]: I don't mind too much either way

**Review by elsirion — APPROVED** [review]

File: fedimint-server/src/net/api.rs [hunk 5]
```diff
@@ -115,10 +115,10 @@ impl ConsensusApi {
 
             let cache = module.build_verification_cache(&[input.clone()]);
             let meta = module
-                .validate_input(
+                .process_input(
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply fedimint-server/src/net/api.rs:120]: Just catching up on this. What is the purpose of being able to write to a db transaction but never commit it?

**Pull request merged**